runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// sync read path
+	err = db.syncRead(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create read handle
+	db.readHandle, err = db.openDBAndAttach(ctx, true)
+	if err != nil {
+		if strings.Contains(err.Error(), "Symbol not found") {
+			fmt.Printf("Your version of macOS is not supported. Please upgrade to the latest major release of macOS. See this link for details: https://support.apple.com/en-in/macos/upgrade")
+			os.Exit(1)
+		}
+		return nil, err
+	}
+
+	return db, nil
+}
+
+type db struct {
+	dbIdentifier string
+	opts         *DBOptions
+
+	readHandle *sqlx.DB
+	readPath   string
+	writePath  string
+	readMu     sync.RWMutex
+	writeMu    sync.Mutex
+	writeDirty bool
+
+	backup *blob.Bucket
+
+	logger *slog.Logger
+}
+
+var _ DB = &db{}
+
+func (d *db) Close() error {
+	d.writeMu.Lock()
+	defer d.writeMu.Unlock()
+
+	d.readMu.Lock()
+	defer d.readMu.Unlock()
+
+	return d.readHandle.Close()
+}
+
+func (d *db) AcquireReadConnection(ctx context.Context) (Conn, func() error, error) {
+	d.readMu.RLock()
+
+	c, err := d.readHandle.Connx(ctx)
+	if err != nil {
+		d.readMu.RUnlock()
+		return nil, nil, err
+	}
+
+	return &conn{
+			Conn: c,
+			db:   d,
+		}, func() error {
+			err = c.Close()
+			d.readMu.RUnlock()
+			return err
+		}, nil
+}
+
+func (d *db) AcquireWriteConnection(ctx context.Context) (Conn, func() error, error) {
+	d.writeMu.Lock()
+	defer d.writeMu.Unlock()
+	c, release, err := d.acquireWriteConn(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &conn{
+		Conn: c,
+		db:   d,
+	}, release, nil
+}
+
+func (d *db) CreateTableAsSelect(ctx context.Context, name, query string, opts *CreateTableOptions) error {
+	if opts == nil {
+		opts = &CreateTableOptions{}
+	}
+	d.logger.Debug("create table", slog.String("name", name), slog.Bool("view", opts.View))
+	d.writeMu.Lock()
+	defer d.writeMu.Unlock()
+	conn, release, err := d.acquireWriteConn(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = release()
+	}()
+	return d.createTableAsSelect(ctx, conn, release, name, query, opts)
+}
+
+func (d *db) createTableAsSelect(ctx context.Context, conn *sqlx.Conn, releaseConn func() error, name, query string, opts *CreateTableOptions) error {
+	// check if some older version exists
+	oldVersion, oldVersionExists, _ := tableVersion(d.writePath, name)
+	d.logger.Debug("old version", slog.String("version", oldVersion), slog.Bool("exists", oldVersionExists))
+
+	// create new version directory
+	newVersion := newVersion()
+	newVersionDir := filepath.Join(d.writePath, name, newVersion)
+	err := os.MkdirAll(newVersionDir, fs.ModePerm)
+	if err != nil {
+		return fmt.Errorf("create: unable to create dir %q: %w", name, err)
+	}
+
+	var m meta
+	if opts.View {
+		// create view - validates that SQL is correct
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s\n)", safeSQLName(name), query))
+		if err != nil {
+			return err
+		}
```

Comment:
I did not understand this.
Here we are creating the view in `writeHandle` to just validate that the SQL is correct. This handle will be closed when function returns.
The `readHandle` is always recreated whenever db is updated and recreates the view from the SQL in `metdata` file.

Reply:
Okay, I think I missed this metadata file logic in the PR. So it will not try to attach databases with views in the read path?

What I mean is if we create a view in one file that relies on a table in another file. And then the table is updated (with the same schema, but new version). Does it continue to work correctly?

Reply:
Okay now I understood the issue. 
Actually in the earlier logic, in write path I was attaching attach all databases to an in-memory db. For creating a table I was attached a new db and then ingested data. For a view I did not create any db but just created a `meta.json` file.
In the read path, for tables, the db will be attached and then I would create views using just the sql from `meta.json`(Though I now realise that this approach is not correct. For this to work correctly we need to create all views in the original sequence otherwise a view on top of a view may not work which means we need more bookkeeping considering renames, deletes and drops).

With the new logic where we attach all db's to the table specific db file we will create a db file with a view and then in-memory db will be another view on top of this view. Since the views are not recreated we don't have the problem as in the previous approach. 
The views continue to work.